### PR TITLE
Iterate #3663

### PR DESCRIPTION
**Pull request by lcaresia**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/iterate/sources/new-question-answer/new-question-answer.mjs [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import iterate from "../../iterate.app.mjs";
+
+export default {
+  name: "New Question Answer",
+  version: "0.0.3",
+  key: "iterate-new-question-answer",
+  description: "Emit new event on a question is answered.",
```
Comment (GTFalcao) [inline components/iterate/sources/new-question-answer/new-question-answer.mjs:7]: Minor suggestion:
```suggestion
  description: "Emit new event when a question is answered.",
```

File: components/iterate/sources/new-survey-response/new-survey-response.mjs [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import iterate from "../../iterate.app.mjs";
+
+export default {
+  name: "New Survey Response",
+  version: "0.0.3",
```
Comment (GTFalcao) [inline components/iterate/sources/new-survey-response/new-survey-response.mjs:5]:
```suggestion
  version: "0.0.1",
```

File: components/iterate/sources/new-question-answer/new-question-answer.mjs [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import iterate from "../../iterate.app.mjs";
+
+export default {
+  name: "New Question Answer",
+  version: "0.0.3",
```
Comment (GTFalcao) [inline components/iterate/sources/new-question-answer/new-question-answer.mjs:5]:
```suggestion
  version: "0.0.1",
```

File: components/iterate/sources/new-question-answer/new-question-answer.mjs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+      if (
+        answers.length < 100 ||
+        answers.filter((answer) => answer.id === lastAnswerId).length
+      ) {
+        return;
+      }
+
+      nextUrl = links?.next ?? null;
+    } while (nextUrl);
```
Comment (GTFalcao) [inline components/iterate/sources/new-question-answer/new-question-answer.mjs:80]: `nextUrl` would only be assigned `null` if it was already null or undefined, in which case it'd already break the loop.
So I think that's not really needed.
```suggestion
      nextUrl = links?.next;
    } while (nextUrl);
```

File: components/iterate/sources/new-survey-response/new-survey-response.mjs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  key: "iterate-new-survey-response",
+  description: "Emit new event on a survey is responded.",
```
Comment (GTFalcao) [inline components/iterate/sources/new-survey-response/new-survey-response.mjs:7]: Minor suggestion:
```suggestion
  description: "Emit new event when a survey is answered.",
```

File: components/iterate/sources/new-survey-response/new-survey-response.mjs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      if (
+        responses.length < 100 ||
+        responses.filter((response) => response.id === lastResponseId).length
+      ) {
+        return;
+      }
+
+      nextUrl = links?.next ?? null;
+    } while (nextUrl);
+  },
```
Comment (GTFalcao) [inline components/iterate/sources/new-survey-response/new-survey-response.mjs:81]: Both sources have nearly identical functionality. The only difference is the app method they use (`getAnswers` vs `getResponses`).

I think we would benefit greatly from a `common` file here to reuse this shared code. You can point to a common method, something such as `getEntries`, and each component would define this method to point to the app's `getResponses` or `getAnswers` depending on which they use.

If the `db` entries have to be named differently, a method for obtaining the db name would also work, while the actual `get` and `set` methods should be in the common code.

**Review by GTFalcao — CHANGES_REQUESTED** [review]
Hi @lcaresia , thanks for your contribution!

The two sources have nearly identical code, so I think a `common` file would be of great value here. I've left a few comments regarding this and some other small suggestions.

**Review by GTFalcao — APPROVED** [review]
Looks good to me! 👏 
Moving to QA

**Pull request merged**